[mbx] Update mailbox documentation (#27194)
lowRISC/opentitan: hw/ip/mbx/doc/programmers_guide.md

# Programmer's Guide

In a typical use case there are two software agents involved in the mailbox communication:

- RoT firmware configures the mailbox IP block and awaits requests from the SoC side.
- SoC-side software makes requests of the RoT firmware and awaits its responses.

> A note on terminology: The request is received from the SoC side into an _Inbox_ on the RoT and the response is collected from the _Outbox_, i.e. the mailbox directions are named from the perspective of the RoT.

Author: Thanks. That is clearer.

Each of the request and the response is composed of 'DWORDs' which are 32 bits in size, so all messages sent and received by the mailbox must be a multiple of 4 bytes in length.
An additional constraint of at most 1K DWORDs (4KiB) applies to the size of any response from the RoT (Outbox message).

## Initialization

Since the mailbox logic acts as a host on the TileLink fabric, before requests may be received from the SoC side the firmware of the RoT must configure the address ranges that may be accessed by the Inbox (SoC->RoT) and Outbox (RoT->SoC) logic.

> To configure each of the Inbox and Outbox sides two registers must be programmed with a suitable address range:

Reviewer: ```suggestion
Four registers -- two for each side (Inbox and Outbox) -- must be programmed with a suitable address range:
```

Author: Reworded.

## Request message from SoC to RoT

A software/hardware agent on the SoC side is required to check that the BUSY bit is clear before starting to send a request to the RoT.
The BUSY bit indicates that the mailbox is already part-way through a request-response operation, so another request cannot be accepted.

The request itself is sent by writing each 32-bit DWORD of the message to the WDATA register on the SoC-side interface.
The write data will be accepted into the Inbox handler when it can be presented onto the RoT-side fabric for writing to the Inbox memory.
A back-pressure/flow control mechanism prevents data loss at this point in the event that the SoC side faster.

The request is written into the mailbox memory of the RoT side starting at the address specified in [`inbound_base_address`](registers.md#inbound_base_address) and incrementing for each DWORD transferred.
When the entire request has been written to the Inbox via the WDATA register, the SoC-side agent must set the GO bit of the [`soc_control`](registers.md#soc_control) register to present it to the RoT firmware.

> The causes the Inbox handler to assert the `mbx_ready` to the RoT firmware so that it may collect and process the request message.

Reviewer: ```suggestion
This causes the Inbox handler to assert the `mbx_ready` to the RoT firmware so that the firmware may collect and process the request message.
```

The RoT firmware shall consult the [`inbound_write_ptr`](registers.md#inbound_write_ptr) register to ascertain the length of the transferred request message.
The address in this register is advanced whenever a DWORD of the request message is written to the Inbox, so upon receipt of the complete request it holds the 4-byte aligned address _beyond_ the final DWORD of the request. The register [`inbound_write_ptr`](registers.md#inbound_write_ptr) is a read only register to the RoT firmware and its value is automatically reset to that of [`inbound_base_address`](registers.md#inbound_base_address) at the start of each request message.

## Collection of RoT response message by SoC

When the RoT firmware has performed the requested operation it is required to produce a response for the SoC side to collect.
The firmware shall write the response message into the Outbox memory starting at the address specified in [`outbound_base_address`](registers.md#outbound_base_address), and shall then specify the number of DWORDs of message data by writing a non-zero value to the [`outbound_object_size`](registers.md#outbound_object_size) register.

> This register permits the sending of responses that are up to 1K DWORDs in length for a maximum size of 4KiB.

Reviewer: ```suggestion
This register permits the sending of responses that are up to 1024 DWORDs in length for a maximum size of 4 KiB.
```

Author: Changed.

## Abort mechanism

> If the SoC side needs to cancel an in-progress service request it may use the Abort mechanism of the DOE mailbox control register.

Reviewer: ```suggestion
If the SoC side needs to cancel an in-progress service request, it may use the Abort mechanism of the DOE mailbox control register.
```
